### ISSUE TITLE
possible STM32F4 board candidate

### ISSUE THREAD

**Issue by extrowerk** [chettrick/discobsd#15]
There is a new (?) cheap board family available via aliexpress, which could be a nice candidate for porting.

It is sold under the name "WeAct STM32F405RGT6 STM32F412RET6 STM32F446RET6 STM32F405 STM32F412 STM32F446 STM32F4 STM32 Core Board Demo Board Micropython"  and it is sold under US$5.

Please consider to port DiscoBSD to this board.

Specs:

STM32F412RET6

STM32 Dynamic Efficiency MCU with BAM, High-performance and DSP with FPU, Arm Cortex-M4 MCU with 512 Kbytes of Flash memory, 100 MHz CPU, Art Accelerator, DFSDM

STM32F446RET6

High-performance foundation line, Arm Cortex-M4 core with DSP and FPU, 512 Kbytes of Flash memory, 180 MHz CPU, ART Accelerator, Dual QSPI

STM32F405RGT6

High-performance foundation line, Arm Cortex-M4 core with DSP and FPU, 1 Mbyte of Flash memory, 168 MHz CPU, ART Accelerator

DataLink

https://github.com/WeActStudio/WeActStudio.STM32F4_64Pin_CoreBoard.git

-----

There are other alternative boards from the same company, the "WeAct STM32H723VGT6 STM32H723 STM32H7 STM32 Core Board Demo Board" looks pretty nice aswell

**Comment by extrowerk** [comment]
Or the "WeAct STM32H7 STM32H743 STM32H743VIT6 STM32 Board 2M Flash 1M RAM Learning Board Development board"

**Comment by chettrick** [comment]
@extrowerk Those boards look interesting. I wonder how they can be priced below the unit price for an STM32F4 part. Perhaps the chips are clones?
Even if they are clones, and they are largely similar to the regular STM32F4, then I don't see much issue in supporting these boards.
I do not yet have any of these boards but I will look into getting some soon.
Thanks for the pointer.

- Chris

Also, patches are appreciated, if you get to porting to them before I do.

**Comment by extrowerk** [comment]
Sadly i am not up for the task, but let us know which model you pick, so we can do beta-tests :)

**Comment by chettrick** [comment]
I ordered each of the boards with STM32F412, STM32F446, and STM32F405 chips, but it will take a while before they are shipped to me.

DiscoBSD already supports the F412 and F407, so the WeAct boards with those chips will be easy to port.
I will port the NUCLEO-F446RE board before starting on the WeAct F446 board.

A bit of work is required to support the STM32F7 and H7 chips (Cortex-M7), so I'll currently be skipping boards with those chips in the near future.

Also, I have done a bit of work in supporting Cortex-M0+, which will be good for the obvious target of the Raspberry Pi Pico RP2040. More work still needs to be done, though.

**Comment by chettrick** [comment]
Commit 3488639 enables the building and cleaning of the sys/stm32/f412weactcore kernel compile directory.

**Comment by extrowerk** [comment]
Nice! Which other models do you plan to support if any?

**Comment by extrowerk** [comment]
I have currently only an OX64, an Rpi Pico and similar unsupported boards.
I am interested in the smallest viable system for a unixy handheld with a VT100-like display and a keyboard. Possibly with some kind of networking, ideally.

**Comment by chettrick** [comment]
I have just added initial support for the WeAct Studio STM32F412RET6 Core board in commit 493325861c47960aa73d45c5c721dd5ae438f591.
Review the commit message for all the specific details.

These boards are pretty good, especially their small size. Major downsides are that they don't include an onboard debugger chip like an ST-Link nor a virtual com port. A USB to serial FTDI cable with female pins can be used to connect to the debug header for the standard console access to DiscoBSD. That works pretty good. An just use a spare Nucleo board for its ST-Link for debugging.

A USB uart would be nice to have, but the STM32Cube USB driver software in the HAL has a separate license that is not appropriate for inclusion in this operating system. Any pointers to better licensed USB uart for STM32 would be great.

**Comment by chettrick** [comment]
> I have currently only an OX64, an Rpi Pico and similar unsupported boards. I am interested in the smallest viable system for a unixy handheld with a VT100-like display and a keyboard. Possibly with some kind of networking and USB support, ideally.

Have you looked at the [Beepy][1] yet?
I would like to get DiscoBSD running on that form factor.

[1]: https://beepy.sqfmi.com/

**Comment by extrowerk** [comment]
Hi, thanks for the hint, i know about Beepy, but it is not available at tihs moment sadly :(
I also like the BlackBerry-liek form-factor...

**Comment by extrowerk** [comment]
I am happy to see your nice progress! The community now have access to cheap, small and available boards to test BSD on them. :)

**Comment by chettrick** [comment]
Yes it is a shame about the lack of availability for a Beepy.

Thank you for the kind words! I started this whole project to create something I wanted to exist in the world, and no one had done it yet. I searched and searched for any actual progress and found none. The closest I found was [this post on the old RetroBSD forum][1] (before it was shut down) for people wanting to have RetroBSD on Arm, but no one rose to the challenge and made it happen. So I did.

For a long while now, I have completed all the required porting tasks outlined by Serge for a port of RetroBSD to another architecture. Only one item remains, and that is using the memory protection unit to guard kernel data. I will be working on that once I finish setting up a continuous integration build server for automatically doing full snapshots and releases.

[1]: https://web.archive.org/web/20150919163519/http://retrobsd.org/viewtopic.php?f=3&t=2545

**Comment by chettrick** [comment]
This is the list I wrote down from Serge V's post in that forum thread:

```
List of items to be done for ARM port, from Serge V:
- Most of kernel is arch-independent (sys/kernel,
  sys/include) ==> There is a small number of places where a
  saved stack frame is directly referenced (look for
  u.u_frame), it needs #ifdef guards.
- Make sys/stm32 like sys/pic32.
- Use sys/pic32/baremetal/Makefile config as starting point.
- Make copies of arch-dependent files from sys/pic32 to
  sys/stm32 and modify them:
  cpu.h, io.h, limits.h, machparam.h, startup.S, clock.c,
  cons.c, devsw.c, exception.c, machdep.c, mem.c, signal.c,
  sysctl.c.
- Design memory map and linker script
- Design user/kernel API: how to pass args to syscalls and
  get returned data. ==> Build syscall library at
  src/libc/thumb2/sys (maybe arm instead of thumb2).
- Port assembler files from NetBSD or OpenBSD.
- MPU used to protect kernel data region from user programs.
```

Of course there are a lot more areas that need to be addressed (for example, signal handling), but it was a good starting point.

**Comment by chettrick** [comment]
> The community now have access to cheap, small and available boards to test BSD on them. :)

Even on Windows it is very easy to run a recent version of DiscoBSD.

[Olimex PIC32-RETROBSD][1] board for running **DiscoBSD/pic32**:
* Use [pic32prog][2] to flash a new kernel.
* Use [Rufus][3] to image the pic32 file system to a MicroSD card.

[WeAct STM32F4 board][4] for running **DiscoBSD/stm32**:
* Use [STM32CubeProgrammer][5] to flash a new kernel.
* Use [Rufus][3] to image the stm32 file system to a MicroSD card.

Can't get cheaper and easier than that.

[1]: https://www.olimex.com/Products/Duino/PIC32/PIC32-RETROBSD/
[2]: https://github.com/sergev/pic32prog
[3]: https://rufus.ie/
[4]: https://www.aliexpress.com/item/1005006051651136.html
[5]: https://www.st.com/en/development-tools/stm32cubeprog.html

**Comment by chettrick** [comment]
That is quite the beast of a processor! 1M of ram. 
It would be helpful to have loads of ram in the future for running compilers. 
That chip is a Cortex-M7 part, so it is a bit different than the Cortex-M4 that is currently targeted by DiscoBSD. It wouldn't be a ton of work to do the port though.

**Comment by extrowerk** [comment]
We have also this: https://www.aliexpress.com/item/1005002953507348.html
and this: 
https://www.aliexpress.com/item/1005004632489443.html

**Comment by extrowerk** [comment]
I am looking at the WeAct STM32F405RGT6 STM32F412RET6 STM32F446RET6 boards, they became pretty cheap i nthe meantime (~2000 HUF, eg. 5 EUR or so), but before I buy anything I would like to hear your opinion, which one is the recommended? If none of them, please recommend me some other board.
Thank You!

**Comment by chettrick** [comment]
The 405 and 412 WeAct boards are currently supported and they work well. The 446 board will be supported in the future.
The main issues with the WeAct boards are that the debug pin header needs to be soldered onto the board, and that they do not have a built-in ST-Link programmer with a usb serial com port.
A spare ST Nucleo board or STLink clone can be used for that though.
AliExpress also has a "[Mini Debugger STLink][1]" for pretty cheap. I have not used one yet, but they look like a nice pairing with a WeAct board such as the 412 board.
The [STLINK-V3MINIE][2] is good and inexpensive, but needs a header soldered on and then use jumper wires to the WeAct board for debugging and serial port.

My main development board is the [STM32F412G-DISCO][3] from STMicro, and it is the default BOARD in DiscoBSD.
I like it, as it is the cheapest STM32F4 development board that includes a MicroSD slot (wired for fast SDIO) and has 256Kbytes of RAM. It is large in physical size, though. It works great out of the box. No soldering or jumper wires needed.

Let me know what you choose to do. I will like to support as best as I can.

[1]: https://www.aliexpress.com/item/1005005303809188.html
[2]: https://www.st.com/en/development-tools/stlink-v3minie.html
[3]: https://www.st.com/en/evaluation-tools/32f412gdiscovery.html

**Comment by chettrick** [comment]
All three WeAct Studio STM32F4 Core Boards are now supported in DiscoBSD/stm32.

Kernels are F405WEACTCORE, F412WEACTCORE, and F446WEACTCORE, available in sys/arch/stm32/compile.

See https://github.com/WeActStudio/WeActStudio.STM32F4_64Pin_CoreBoard for board information.
